hbase-server/src/main/java/org/apache/hadoop/hbase/master/procedure/SnapshotProcedure.java

```diff
@@ -0,0 +1,471 @@
+      List<RegionInfo> regions, String threadPoolName) throws IOException {
+    ThreadPoolExecutor exec = SnapshotManifest
+      .createExecutor(env.getMasterConfiguration(), threadPoolName);
+    try {
+      ModifyRegionUtils.editRegions(exec, regions, new ModifyRegionUtils.RegionEditTask() {
+        @Override
+        public void editRegion(final RegionInfo region) throws IOException {
+          snapshotManifest.addRegion(CommonFSUtils.getTableDir(rootDir, snapshotTable), region);
+          LOG.info("take snapshot region={}, table={}", region, snapshotTable);
+        }
+      });
+    } finally {
+      exec.shutdown();
+    }
+    status.setStatus("Completed referencing closed/split regions of table: " + snapshotTable);
+  }
+
+  private void snapshotMobRegion(MasterProcedureEnv env) throws IOException {
+    if (!MobUtils.hasMobColumns(htd)) {
+      return;
+    }
+    ThreadPoolExecutor exec = SnapshotManifest
+      .createExecutor(env.getMasterConfiguration(), "MobRegionSnapshotPool");
+    RegionInfo mobRegionInfo = MobUtils.getMobRegionInfo(htd.getTableName());
+    try {
+      ModifyRegionUtils.editRegions(exec, Collections.singleton(mobRegionInfo),
+          new ModifyRegionUtils.RegionEditTask() {
+          @Override
+          public void editRegion(final RegionInfo region) throws IOException {
+            snapshotManifest.addRegion(CommonFSUtils.getTableDir(rootDir, snapshotTable), region);
+          }
+        });
+    } finally {
+      exec.shutdown();
+    }
+    status.setStatus("Completed referencing HFiles for the mob region of table: " + snapshotTable);
+  }
+
+  @Override
+  protected void serializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.serializeStateData(serializer);
+    serializer.serialize(SnapshotProcedureStateData
+      .newBuilder().setSnapshot(this.snapshot).build());
+  }
+
+  @Override
+  protected void deserializeStateData(ProcedureStateSerializer serializer) throws IOException {
+    super.deserializeStateData(serializer);
+    SnapshotProcedureStateData data = serializer.deserialize(SnapshotProcedureStateData.class);
+    this.snapshot = data.getSnapshot();
+  }
+
+  private Procedure<MasterProcedureEnv>[] createRemoteSnapshotProcedures(MasterProcedureEnv env) {
+    return env.getAssignmentManager().getTableRegions(snapshotTable, true)
+      .stream().filter(r -> RegionReplicaUtil.isDefaultReplica(r))
+      .map(r -> new SnapshotRegionProcedure(snapshot, r))
+      .toArray(SnapshotRegionProcedure[]::new);
+  }
+
+  @Override
+  public void toStringClassDetails(StringBuilder builder) {
+    builder.append(getClass().getName())
+      .append(", id=").append(getProcId())
+      .append(", snapshot=").append(ClientSnapshotDescriptionUtils.toString(snapshot));
+  }
+
+  public SnapshotDescription getSnapshotDesc() {
+    return snapshot;
+  }
+
+  @Override
+  protected void afterReplay(MasterProcedureEnv env) {
+    try {
+      prepareSnapshotEnv(env);
```

Comment:
Better add a check on the state? If the procedure is at the first state, I do not think we need to call this? We will call this method when executing the procedure.

Reply:
Ok, I will address it. Thanks, Duo.